### PR TITLE
cedar-go: run integration tests as part of the regular test suite

### PR DESCRIPTION
**Pull request by patjakdev**
*Issue #, if available:* None

*Description of changes:*

This change adds a local copy of the corpus-tests.tar.gz from the upstread cedar-policy/cedar-integration-tests repo and a test harness which loads the tests in-memory and runs them.

This will allow us to run these tests as part of our continuous integration pipeline and should reduce noise due to network blips. It will also mean that upstream changes to the formatting of the tests won't instantly break the test runner, leaving developers unsure if their code is correct or not.

The tests run in just over 1 second on my dev machine, so I see no harm in running them on every PR.

Additionally, the nightly test corpus run has been transformed into a check for a difference in the upstream version and the local version of the tests. If a change happens upstream, an issue will be opened which we can use to track updating the local copy of the tests.

### CODE REVIEW

File: corpus_test.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	}
+	defer corpusArchive.Close()
+
+	gzipReader, err := gzip.NewReader(corpusArchive)
+	if err != nil {
+		t.Fatal("error reading corpus compressed archive header", err)
+	}
+	defer gzipReader.Close()
+
+	buf, err := io.ReadAll(gzipReader)
```
Comment (patjakdev) [inline corpus_test.go:97]: `buf` ends up being about 14MB in size with the current version of the integration tests. Not great, not terrible.

File: corpus_test.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		return nil, err
+	}
+
+	return content, nil
+}
+
+func TestCorpus(t *testing.T) {
+	t.Parallel()
+
+	corpusArchive, err := os.Open("corpus-tests.tar.gz")
```
Comment (philhassey) [inline corpus_test.go:85]: If this is referencing the file that is checked into the repo, I'd recommend using `embed` to embed it into the test binary, that way we don't have to worry about what the CWD, etc is.

File: corpus_test.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+				t.Fatal("error reading policy content", err)
+			}
+
+			policySet, err := NewPolicySet("policy.cedar", policyContent)
+			if err != nil {
+				t.Fatal("error parsing policy set", err)
+			}
+
+			for _, request := range tt.Requests {
+				t.Run(request.Desc, func(t *testing.T) {
```
Comment (philhassey) [inline corpus_test.go:155]: I'm not sure if the Go compiler will magically capture request or not in this case. Maybe check?
`request := request` was previously used when firing off goroutines.  This isn't required anymore, but I don't know if lambdas capture it or not.

File: corpus_test.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		Errors    []string   `json:"errors"`
+	} `json:"requests"`
+}
+
+type FileDataPointer struct {
+	Position int64
+	Size     int64
+}
+
+type FileDataMap struct {
```
Comment (philhassey) [inline corpus_test.go:55]: I'd probably call this `tarMap` or something, but it doesn't really matter.

File: corpus_test.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		files: make(map[string]FileDataPointer),
+	}
+}
+
+func (fdm *FileDataMap) AddFileData(path string, position int64, size int64) {
+	fdm.files[path] = FileDataPointer{position, size}
+}
+
+func (fdm FileDataMap) GetFileData(path string) ([]byte, error) {
+	fdp := fdm.files[path]
```
Comment (philhassey) [inline corpus_test.go:72]: A `fdp,ok :=` might be nice with an error if it can't find the file.

**Review by philhassey — APPROVED** [review]
Of my comments the `embed` one is the one I feel the most strongly about.  The goroutine / concern can either be handled by just doing it, or by validating if lambdas auto-capture the loop variable or not.

File: corpus_test.go (hunk 2 above)
Comment (patjakdev) [reply corpus_test.go]: Cool. I saw that it was used before but didn't quite grasp the value. I've reintroduced it now, thanks!

File: corpus_test.go (hunk 3 above)
Comment (patjakdev) [reply corpus_test.go:152]: Yes, [this is mentioned](https://arc.net/l/quote/jlqattmy) in the blog post introducing the change to loop variables.

We can confirm that it's true in our case by printing out test spew with `go test . -v` and observing that the tests have unique names. For example:

```
    --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_4 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_6 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_2 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_3 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_1 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_0 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_7 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_5 (0.00s)
    --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_7 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_6 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_5 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_2 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_3 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_4 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_0 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_1 (0.00s)
```

**Pull request merged**